src/include/common/object_pool.h

```diff
@@ -95,15 +178,38 @@ class ObjectPool {
    * @param obj pointer to object to release
    */
   void Release(T *obj) {
-    if (reuse_queue_.UnsafeSize() > reuse_limit_)
+    if (reuse_queue_.UnsafeSize() >= reuse_limit_) {
       alloc_.Delete(obj);
-    else
+      current_size_--;
+    } else {
       reuse_queue_.Enqueue(std::move(obj));
+    }
+  }
+
+  /***
+   * Check if current_size_ is always no greater than size_limit_
+   *
+   * @return true if current size <= size_limit; false otherwise.
+   */
+  bool CheckInvariance() {
```

Comment:
This should not be a public method if you only expect to check them in tests.

It probably qualifies more as asserts in the functions.